File: src/main/java/org/taktik/icure/logic/impl/filter/healthelement/HealthElementByHcPartyTagCodeFilter.java
package org.taktik.icure.logic.impl.filter.healthelement;

import org.springframework.beans.factory.annotation.Autowired;
import org.taktik.icure.entities.HealthElement;
import org.taktik.icure.entities.User;
import org.taktik.icure.logic.HealthElementLogic;
import org.taktik.icure.logic.ICureSessionLogic;
import org.taktik.icure.logic.impl.filter.Filter;
import org.taktik.icure.logic.impl.filter.Filters;

import javax.security.auth.login.LoginException;
import java.util.HashSet;
import java.util.Set;

public class HealthElementByHcPartyTagCodeFilter implements Filter<String, HealthElement, org.taktik.icure.dto.filter.healthelement.HealthElementByHcPartyTagCodeFilter> {
	HealthElementLogic healthElementLogic;
	ICureSessionLogic sessionLogic;

	@Autowired
	public void setHealthElementLogic(HealthElementLogic healthElementLogic) {
		this.healthElementLogic = healthElementLogic;
	}

	@Autowired
	public void setSessionLogic(ICureSessionLogic sessionLogic) {
		this.sessionLogic = sessionLogic;
	}

	@Override
	public Set<String> resolve(org.taktik.icure.dto.filter.healthelement.HealthElementByHcPartyTagCodeFilter filter, Filters context) {
		try {
			String hcPartyId = filter.getHealthCarePartyId() != null ? filter.getHealthCarePartyId() : getLoggedHealthCarePartyId();
			HashSet<String> ids = null;
			if (filter.getTagType() != null && filter.getTagCode() != null) {
				ids = new HashSet<>(healthElementLogic.findByHCPartyAndTags(hcPartyId, filter.getTagType(), filter.getTagCode()));
				;
			}

            if (filter.getCodeType() != null && filter.getCodeNumber() != null) {
				HashSet<String> byCode = new HashSet<>(healthElementLogic.findByHCPartyAndCodes(hcPartyId, filter.getCodeType(), filter.getCodeNumber()));
				if (ids == null) {
					ids = byCode;
				} else {
					ids.retainAll(byCode);
				}
			}

			if (filter.getStatus() != null) {
				HashSet<String> byStatus = new HashSet<>(healthElementLogic.findByHCPartyAndStatus(hcPartyId, filter.getStatus()));

				if (ids == null) {
					ids = byStatus;
				} else {
					ids.retainAll(byStatus);
				}
			}

			return ids != null ? ids : new HashSet<>();
		} catch (LoginException e) {
			throw new IllegalArgumentException(e);
		}
	}

	private String getLoggedHealthCarePartyId() throws LoginException {
		User user = sessionLogic.getCurrentSessionContext().getUser();
		if (user == null || user.getHealthcarePartyId() == null) {
			throw new LoginException("You must be logged to perform this action. ");
		}
		return user.getHealthcarePartyId();
	}
}

File: src/main/kotlin/org/taktik/icure/asynclogic/impl/filter/healthelement/HealthElementByHcPartyTagCodeFilter.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */

package org.taktik.icure.asynclogic.impl.filter.healthelement

import javax.security.auth.login.LoginException
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.toSet
import org.springframework.stereotype.Service
import org.taktik.icure.asynclogic.AsyncSessionLogic
import org.taktik.icure.asynclogic.HealthElementLogic
import org.taktik.icure.asynclogic.impl.filter.Filter
import org.taktik.icure.asynclogic.impl.filter.Filters
import org.taktik.icure.domain.filter.healthelement.HealthElementByHcPartyTagCodeFilter
import org.taktik.icure.entities.HealthElement
import org.taktik.icure.utils.getLoggedHealthCarePartyId

@Service
class HealthElementByHcPartyTagCodeFilter(private val healthElementLogic: HealthElementLogic,
                                          private val sessionLogic: AsyncSessionLogic) : Filter<String, HealthElement, HealthElementByHcPartyTagCodeFilter> {

    override fun resolve(filter: HealthElementByHcPartyTagCodeFilter, context: Filters) = flow<String> {
        try {
            val hcPartyId: String = filter.healthCarePartyId ?: getLoggedHealthCarePartyId(sessionLogic)
            var ids: HashSet<String>? = null
            if (filter.tagType != null && filter.tagCode != null) {
                ids = HashSet(healthElementLogic.listHealthElementIdsByHcPartyAndTags(hcPartyId, filter.tagType!!, filter.tagCode!!).toSet())
            }
            if (filter.codeType != null && filter.codeNumber != null) {
                val byCode = HashSet(healthElementLogic.listHealthElementIdsByHcPartyAndCodes(hcPartyId, filter.codeType!!, filter.codeNumber!!).toSet())
                if (ids == null) {
                    ids = byCode
                } else {
                    ids.retainAll(byCode)
                }
            }
            if (filter.status != null) {
                val byStatus = HashSet(healthElementLogic.listHealthElementIdsByHcPartyAndStatus(hcPartyId, filter.status!!).toSet())
                if (ids == null) {
                    ids = byStatus
                } else {
                    ids.retainAll(byStatus)
                }
            }
            ids?.forEach { emit(it) }
        } catch (e: LoginException) {
            throw IllegalArgumentException(e)
        }
    }
}
